packages/plugin-ext/src/api/rpc-protocol.ts

```diff
@@ -301,6 +303,68 @@ class MessageFactory {
     }
 }
 
+/**
+ * This functions are responsible for correct transferring objects via rpc channel.
+ *
+ * To reach that some specific kind of objects is converteed to json in some custom way
+ * and then, after receiving, revived to objects again,
+ * so there is feeling that object was transferred via rpc channel.
+ *
+ * To distinguish between regular and altered objects, field $type is added to altered ones.
+ * Also value of that field specifies kind of the object.
+ */
+namespace ObjectsTransferrer {
+
+    // tslint:disable-next-line:no-any
+    export function replacer(key: string | undefined, value: any): any {
+        if (value instanceof URI) {
```

Comment:
should object transformers be pluggable instead of hardcoding them ?

Reply:
I am not sure we need pluggable architecture here right now. We alter only objects of two types.
Storing that simple things into separate classes (and creating bindings) will make code less clear for current quantity of transferred objects. To me, we should make it pluggable when the quantity of objects will be bigger, because for now it is not clear for me how many and how often this mechanism will be used. But it is only my opinion. WDYT?

Reply:
@mmorhun plug-in system is extensible so ppl would have to make this kind of stuff as well. Without extensibility it will be impossible to have custom serialization.

Reply:
@benoitf do we plan to allow for plugin writers changing of rpc transfer mechanism? It is not safe.

Reply:
not plugin writers, extension that bring plug-ins 

for example plugin-ext bring support of `@theia/plugin` but you would want some serializers for `@che/plugin` or `@customnamespace/plugin`, etc

Reply:
Makes sense, thank you. I'll rework that part.

Reply:
It turned out, that this is not trivial task as it seemed to me before I started work on it.
I've created [issue](https://github.com/theia-ide/theia/issues/4090) for this task.